### DISCUSSION TITLE
Cannot read properties of undefined using a getter accessing outer scope variable

### DISCUSSION THREAD

**Discussion by emavitta** [alpinejs/alpine#3276] [7. Bugs (archived)]
I'm trying to use a getter that have to access an outer scope variable (as seen in this pr #2128)
It seems to work fine but not with nested data. Unless i print the getter in the page with an x-text, i get an "undefined" error, like in this example

````
<div
      x-data="{ foo: 'bar', test:{ data:0 }
    }"
    >
      <div
        x-data="{
        get bob() {
            return this.test.data
        },

    }"
      >
        <button @click="console.log(bob)">alert</button>
      </div>
    </div>
````

The alert will print the correct number, but the console is giving me some error.

![image](https://user-images.githubusercontent.com/57635138/203030134-cbdb3bae-47cc-46ff-aa05-0337d3abe797.png)

Using the getter for something that doesn't need to be printed but it has to be used later in the code doesn't work, like this example

````
<div
      x-data="{ foo: 'bar', test:{ data:0 }
    }"
    >
      <div
        x-data="{
          test:'',
        get bob() {
            return this.test.data
        },
        async fetchCall() {
          const response = await fetch('https://jsonplaceholder.typicode.com/todos/'+this.bob)
          const data = await response.json()
          this.test = data
        }
    }"
      >
        <div x-text="test"></div>
        <button @click="fetchCall">api call</button>
      </div>
    </div>
```

**Comment by ekwoka** [comment]
Well, the second one doesn't work because it has its own `test` prop defined.

The first is quite curious since that should work...

What if you remove the `this`? Since `test` is defined in the scope of that declaration already.

**Comment by SimoTod** [comment]
https://codepen.io/SimoTod/pen/vYrpqgw?editors=1111 it seems to work.
Which version of Alpine are you using? What is that backend.js reference in the error?

**Reply by yabdab** [reply]
@SimoTod I am having similar, if not exact issue, when trying to access a parent property inside the getter. My code looks like you have it in your example, but if keeps saying the parent item is "undefined".

The code looks like this ...

```js
get valid() {
 var found = this.formItems.find((match) => match.name == this.equalto_name);
        if (found) {
          if (this.value != found.value) {
            return false;
          }
        }
 return true;
}
```

Not sure where to look next?

**Reply by SimoTod** [reply]
can you post a full runnable example including the parent data?

**Reply by yabdab** [reply]
I will see what I can do. My app with quite complex. A simple example like you shared works well  in my testing , but my app has the parent item is undefined  no matter what I try. Back to do more debugging 😖

**Reply by yabdab** [reply]
@SimoTod  OK, I wrote a real simple example the exhibits the issue I am having.

```html
<div x-data="myApp">
      <!-- add to appItems each time component is included -->
      <div x-data="myComponent" x-init="addItemToApp($el)">
        <span x-text="valid"></span>
      </div> 
      <button @click.prevent="submitApp">Submit</button>
    </div>
```

and the js ...

```js
   document.addEventListener("alpine:init", () => {
        Alpine.data("myApp", () => ({
          property: "my value",
          appItems: [],
          addItemToApp(el) {
            let item = Alpine.closestDataStack(el)[0];
            this.appItems.push(item);
          },
          init(){
            this.$watch("appItems", (value) =>  console.debug("watching") );
          },
          submitApp(event){
            this.appItems.forEach((item) => {
              console.debug(item.valid);
            });
          }
        }));
        Alpine.data("myComponent", () => ({
          value1: "v1",
          value2: "v2",
          init(){
            // appItems can be seen here, but not below in the getter
            console.debug(this.appItems);
          },
          get valid() {
            // appItems will be undefined when executing submitApp in parent 
            console.debug(this.appItems);
            return true;
          },
        }));
      });
```

When clicking the button, you will see in console that the appItems array is `undefined`, BUT is viewable on the component `init()`.

to test live I set this up for you...

https://yabdab.software/alpine.html

**Reply by SimoTod** [reply]
I tried to console log this and, inside the getter, `this` is just a plain object not one of the Alpine proxies that can access the parent scope magically so I don't think that logic can work.

**Reply by yabdab** [reply]
What is different about your example and mine?  

If it can't be done , then I can accept that. Just wondering 🤷‍♂️

Would it be better to use $store for the appItems instead? That way I can. access it globally.

**Reply by SimoTod** [reply]
I'll have a look tomorrow. It might be something to do with watch

**Reply by ekwoka** [reply]
The issue here is that you are only putting into the appItems array that single data context, not the whole datastack.

If you change it to `Alpine.$data(el)` you would get the merged data stack.

[this sandbox demonstrates this](https://awesomealpine.com/play/?coreplugins=0&html=PGRpdiB4LWRhdGE9Im15QXBwIj4KICA8ZGl2CiAgICB4LWRhdGE9Im15Q29tcG9uZW50IgogICAgeC1pbml0PSJhZGRJdGVtVG9BcHAoJGVsKSIKICA-CiAgICA8c3BhbiB4LXRleHQ9InZhbGlkIj48L3NwYW4-CiAgPC9kaXY-CiAgPGJ1dHRvbiBAY2xpY2sucHJldmVudD0ic3VibWl0QXBwIj5TdWJtaXQ8L2J1dHRvbj4KPC9kaXY-Cg&tw=1&script=QWxwaW5lLmRhdGEoJ215QXBwJywgKCkgPT4gKHsKICBwcm9wZXJ0eTogJ215IHZhbHVlJywKICBhcHBJdGVtczogW10sCiAgYWRkSXRlbVRvQXBwKGVsKSB7CiAgICBsZXQgaXRlbSA9IEFscGluZS4kZGF0YShlbCk7CiAgICB0aGlzLmFwcEl0ZW1zLnB1c2goaXRlbSk7CiAgfSwKICBpbml0KCkgewogICAgQWxwaW5lLmVmZmVjdCgoKSA9PgogICAgICB0aGlzLmFwcEl0ZW1zLmZvckVhY2goCiAgICAgICAgY29uc29sZS5sb2cuYmluZChudWxsLCAnaW4gaW5pdCBlZmZlY3QnKSwKICAgICAgKSwKICAgICk7CiAgfSwKICBzdWJtaXRBcHAoZXZlbnQpIHsKICAgIHRoaXMuYXBwSXRlbXMuZm9yRWFjaCgoaXRlbSkgPT4gewogICAgICBjb25zb2xlLmRlYnVnKCdpbiBzdWJtaXQgbG9vcCcsIGl0ZW0pOwogICAgfSk7CiAgfSwKfSkpOwpBbHBpbmUuZGF0YSgnbXlDb21wb25lbnQnLCAoKSA9PiAoewogIHZhbHVlMTogJ3YxJywKICB2YWx1ZTI6ICd2MicsCiAgaW5pdCgpIHsKICAgIC8vIGFwcEl0ZW1zIGNhbiBiZSBzZWVuIGhlcmUsIGJ1dCBub3QgYmVsb3cgaW4gdGhlIGdldHRlcgogICAgY29uc29sZS5kZWJ1ZygnaW4gaW5pdCcsIHRoaXMuYXBwSXRlbXMpOwogIH0sCiAgZ2V0IHZhbGlkKCkgewogICAgLy8gYXBwSXRlbXMgd2lsbCBiZSB1bmRlZmluZWQgd2hlbiBleGVjdXRpbmcgc3VibWl0QXBwIGluIHBhcmVudAogICAgY29uc29sZS5kZWJ1ZygnaW4gdmFsaWQnLCB0aGlzLmFwcEl0ZW1zKTsKICAgIHJldHVybiBCb29sZWFuKHRoaXMuYXBwSXRlbXMpOwogIH0sCn0pKTsK&v=3.14.0&panel=script)

You shouldn't be pushing the context into the parent like that though...

**Reply by yabdab** [reply]
@ekwoka  "You shouldn't be pushing the context into the parent like that though..."

Why is that?

This is a drag and drop app that allows the enduser to add "items" to the app by dropping them in. Would you do it differently? Maybe put the items in a store instead?

Also, I like that Alpine playground. How can I register for that?

**Reply by ekwoka** [reply]
You'd do it by having the parent directly process the thing be added and Inject it from there.

**Reply by yabdab** [reply]
Are you saying that the parent should inject the markup for the items itself dynamically?

This solution won't allow for that unfortunately. Imagine this is a web page design app and the items ned to be laid out a certain way. 

I am just worried why I should not be pushing into the parent is a bad thing? #learning

**Comment by emavitta** [comment]
@ekwoka , you're right about the props... it works properly as intended. I'm having a hard time replicating the error in a simple way, cause i'm implementing it n some php backend. I'll try to do it in a better way, still i had to do some bad hack like
```js

init(){
this.local=this.parent
}
get getterAction(){
return this.local.....
}
```
otherwise it doesn't work. Also it seems that on stackblitz and similar i cannot see the error, while it appears locally.
@SimoTod thanks for noticing, cause i didn't pay enough attention... it's from the alpine devtools. Maybe it's just an error from the dev tools but the code itself may work...

**Comment by emavitta** [comment]
yep it seems that the error is fired only when the Alpine devtools is on, otherwise i got no error. Still it doesn't seems to work. I'll try to replicate the structure and see what's happening.

**Comment by emavitta** [comment]
Aaaaaaaaaaand i correct myself again.... it works as intended, simply the devtools gave me an error and i got stuck with it, cause i could not understand what was happening. So it's a devtools problem.
I know in vue sometimes values in the devtools doens't update because vue it's lazy (see [FAQ](https://devtools.vuejs.org/guide/faq.html#the-data-isn-t-updating-in-the-component-inspector)), and maybe it's the same error.

**Reply by SimoTod** [reply]
The alpine devtools repo hasn't been updated in a while (21st Sept 2021, before the PR you mentioned). There are probably a lot of things that aren't supported.

**Comment by emavitta** [comment]
I can confirm is a devtools bug.
Her's my [full example](https://stackblitz.com/edit/js-i882pj?file=index.html); it works as intended but once you open the devtools it gave me an error.
Thanks again for the support.
